File: src/GitHubExplorer.Domain/Results/Result.cs
namespace GitHubExplorer.Domain.Results;

public readonly struct Result<T>
{
    public bool IsSuccess { get; }
    public bool IsFailure => !IsSuccess;

    public T Value => IsSuccess
        ? _value!
        : throw new InvalidOperationException("Cannot access Value on a failed Result.");

    public GitHubError Error => IsFailure
        ? _error
        : throw new InvalidOperationException("Cannot access Error on a successful Result.");

    private readonly T? _value;
    private readonly GitHubError _error;

    private Result(bool isSuccess, T? value, GitHubError error)
    {
        IsSuccess = isSuccess;
        _value = value;
        _error = error;
    }

    public static Result<T> Success(T value) => new(true, value, default);

    public static Result<T> Failure(GitHubError error) => new(false, default, error);

    public Result<TResult> Map<TResult>(Func<T, TResult> map) =>
        IsSuccess
            ? Result<TResult>.Success(map(_value!))
            : Result<TResult>.Failure(_error);

    public Result<TResult> Bind<TResult>(Func<T, Result<TResult>> bind) =>
        IsSuccess
            ? bind(_value!)
            : Result<TResult>.Failure(_error);

    public TResult Match<TResult>(Func<T, TResult> onSuccess, Func<GitHubError, TResult> onFailure) =>
        IsSuccess
            ? onSuccess(_value!)
            : onFailure(_error);

    public Result<T> MapError(GitHubError error) =>
        IsFailure
            ? Failure(error)
            : this;
}

File: src/frontend/src/lib/result.ts
import type { ApiError } from "./types.js";

export type Result<T> =
	| { readonly ok: true; readonly value: T }
	| { readonly ok: false; readonly error: ApiError };

export function ok<T>(value: T): Result<T> {
	return { ok: true, value };
}

export function err(error: ApiError): Result<never> {
	return { ok: false, error };
}

export function map<T, U>(result: Result<T>, fn: (value: T) => U): Result<U> {
	return result.ok ? ok(fn(result.value)) : result;
}

export function mapError(
	result: Result<unknown>,
	fn: (error: ApiError) => ApiError,
): Result<unknown> {
	return result.ok ? result : err(fn(result.error));
}
